pingcap/br
# lightning/restore: fix #812, ensure engines are restored in fixed order

### What problem does this PR solve? 

Fix #812

### What is changed and how it works?

(Split off from #815). 

#812 is caused by iterating `cp.Engines` in `restoreEngines()`, which is a hash map. We read the map into a sorted array and iterate that instead, so engines are always restored in deterministic order.

### Check List 

Tests 

 - Integration test
 - Manual test
     * Import a 1 TB table with checkpoint
     * Kill Lightning
     * Restart Lightning
     * Note that the same engines (0, 1, 2, ...) are used in both steps 1 and 3, instead of randomly chosen ones.

Code changes

Side effects

Related changes

 - Need to cherry-pick to the release branch
    * release-4.0

### Release Note

 - When resuming from checkpoint, Lightning now keeps restoring the original engines instead of starting to process some random new ones (#815).
